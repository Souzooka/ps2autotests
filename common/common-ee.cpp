#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
