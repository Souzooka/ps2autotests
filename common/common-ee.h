#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
