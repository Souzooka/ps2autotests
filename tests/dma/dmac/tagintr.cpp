#include <common-ee.h>
#include <assert.h>
#include <kernel.h>
#include <malloc.h>
#include <string.h>
#include "../dmaregs.h"
#include "../dmasend.h"
#include "../dmatags.h"

static volatile DMA::Channel *const toSPR = DMA::D9;

void testTagIrq(u32 *buf, bool tagIntEnable, bool tagIrq) {
	*DMA::D_STAT = DMA::D_STAT_CIS9;
	
	static const u32 dmaPacketSize = 512;
	DMA::SrcChainPacket srcTag(dmaPacketSize);
	
	srcTag.Reset();
	srcTag.REF(buf, 1, tagIrq);
	srcTag.REF(buf, 1);
	srcTag.END(0);
	
	if(tagIntEnable) {
		toSPR->chcr |= DMA::CHCR_TIE;
	} else {
		toSPR->chcr &= ~DMA::CHCR_TIE;
	}

	u32 cisBefore = *DMA::D_STAT & DMA::D_STAT_CIS9;
	DMA::SendChain(toSPR, srcTag.Raw(), dmaPacketSize);
	u32 cisAfter = *DMA::D_STAT & DMA::D_STAT_CIS9;
	
	u32 tadrOffset = (u8 *)toSPR->tadr - (u8 *)srcTag.Raw();
	
	printf("Tag interrupt %s, Tag IRQ %s:\n", 
		tagIntEnable ? "enabled" : "disabled",
		tagIrq ? "enabled" : "disabled");
	printf("  CHCR(tag): %08x, TADR offset: %08x, CIS(before): %08x, CIS(after): %08x\n", 
		toSPR->chcr & DMA::CHCR_TAG_MASK, tadrOffset, cisBefore, cisAfter);
}

int main(int argc, char *argv[]) {
	printf("-- TEST BEGIN\n");
	
	u32 *buf = (u32 *)memalign(16, 16 * 1024);
	buf[0] = 0x01234567;
	buf[1] = 0x89ABCDEF;
	buf[2] = 0xDEADBEEF;
	buf[3] = 0x1337C0DE;
	
	testTagIrq(buf, true, true);
	testTagIrq(buf, false, true);
	testTagIrq(buf, false, false);
	
	free(buf);
	
	printf("-- TEST END\n");
	return 0;
}
