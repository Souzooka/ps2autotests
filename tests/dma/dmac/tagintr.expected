-- TEST BEGIN
Tag interrupt enabled, Tag IRQ enabled:
  CHCR(tag): b0000000, TADR offset: 00000010, CIS(before): 00000000, CIS(after): 00000200
Tag interrupt enabled, Tag IRQ disabled:
  CHCR(tag): 70000000, TADR offset: 00000020, CIS(before): 00000000, CIS(after): 00000200
Tag interrupt disabled, Tag IRQ enabled:
  CHCR(tag): 70000000, TADR offset: 00000020, CIS(before): 00000000, CIS(after): 00000200
Tag interrupt disabled, Tag IRQ disabled:
  CHCR(tag): 70000000, TADR offset: 00000020, CIS(before): 00000000, CIS(after): 00000200
-- TEST END
